### PR TITLE
Enable label alignment for token classification datasets

### PR DESCRIPTION
**Pull request by lewtun**
This PR extends the `Dataset.align_labels_with_mapping()` method to support alignment of label mappings between datasets and models for token classification (e.g. NER).

Example of usage:

```python
from datasets import load_dataset

ner_ds = load_dataset("conll2003", split="train")
# returns [3, 0, 7, 0, 0, 0, 7, 0, 0]
ner_ds[0]["ner_tags"]
# hypothetical model mapping with O <--> B-LOC
label2id = {
    "B-LOC": "0",
    "B-MISC": "7",
    "B-ORG": "3",
    "B-PER": "1",
    "I-LOC": "6",
    "I-MISC": "8",
    "I-ORG": "4",
    "I-PER": "2",
    "O": "5"
  }
ner_aligned_ds = ner_ds.align_labels_with_mapping(label2id, "ner_tags")
# returns [3, 5, 7, 5, 5, 5, 7, 5, 5]
ner_aligned_ds[0]["ner_tags"]
```

Context: we need this in AutoTrain to automatically align datasets / models during evaluation. cc @abhishekkrthakur

### CODE REVIEW

File: src/datasets/arrow_dataset.py [hunk 1]
```diff
@@ -4408,30 +4408,46 @@ def align_labels_with_mapping(self, label2id: Dict, label_column: str) -> "Datas
...
+            else:
+                dset_label_names = [
+                    [int2str_function(label_id).lower() if label_id is not None else None for label_id in seq]
+                    for seq in batch[label_column]
+                ]
+                batch[label_column] = [
+                    [label2id[label_name] if label_name is not None else None for label_name in seq]
+                    for seq in dset_label_names
+                ]
             return batch
```
Comment (mariosasko) [inline src/datasets/arrow_dataset.py:4443]: I would rewrite this as follows:

```suggestion
        if isinstance(label_feature, ClassLabel):
            def process_label_ids(batch):
                dset_label_names = [
                    int2str_function(label_id).lower() if label_id is not None else None
                    for label_id in batch[label_column]
                ]
                batch[label_column] = [
                    label2id[label_name] if label_name is not None else None for label_name in dset_label_names
                ]
                return batch
        else:
            def process_label_ids(batch):
                dset_label_names = [
                    [int2str_function(label_id).lower() if label_id is not None else None for label_id in seq]
                    for seq in batch[label_column]
                ]
                batch[label_column] = [
                    [label2id[label_name] if label_name is not None else None for label_name in seq]
                    for seq in dset_label_names
                ]
                return batch
```

**Review by mariosasko — COMMENTED** [review]
Two nits. 

PS: It makes sense to also add support for the `Sequence` type to `class_encode_column`. I can work on that.

File: src/datasets/arrow_dataset.py [hunk 2]
```diff
@@ -4408,30 +4408,46 @@ def align_labels_with_mapping(self, label2id: Dict, label_column: str) -> "Datas
             raise ValueError(f"Column ({label_column}) not in table columns ({self._data.column_names}).")
 
         label_feature = self.features[label_column]
-        if not isinstance(label_feature, ClassLabel):
+        if not (isinstance(label_feature, ClassLabel) or isinstance(label_feature, Sequence)):
```
[the remark this answers is not in the document]
Comment (lewtun) [reply src/datasets/arrow_dataset.py:4411]: Nice catch!

File: src/datasets/arrow_dataset.py (hunk 1 above)
Comment (lewtun) [reply src/datasets/arrow_dataset.py:4443]: Indeed, this is more elegant - thanks for the suggestion!

**Comment by lewtun** [comment]
Hmm, not sure why the Windows tests are failing with:

```
Did not find path entry C:\tools\miniconda3\bin
C:\tools\miniconda3\envs\py37\python.exe: No module named pytes
```

File: src/datasets/arrow_dataset.py [hunk 3]
```diff
@@ -4408,30 +4408,54 @@ def align_labels_with_mapping(self, label2id: Dict, label_column: str) -> "Datas
...
 
         label_feature = self.features[label_column]
-        if not isinstance(label_feature, ClassLabel):
+        if not (
+            isinstance(label_feature, ClassLabel)
+            or (isinstance(label_feature, Sequence) and isinstance(label_feature.feature, ClassLabel))
+        ):
             raise ValueError(
-                f"Aligning labels with a mapping is only supported for {ClassLabel.__name__} column, and column {label_feature} is {type(label_feature).__name__}."
+                f"Aligning labels with a mapping is only supported for {ClassLabel.__name__} or {Sequence.__name__} column, and column {label_feature} is of type {type(label_feature).__name__}."
```
Comment (mariosasko) [inline src/datasets/arrow_dataset.py:4416]:
```suggestion
                f"Aligning labels with a mapping is only supported for {ClassLabel.__name__} column or {Sequence.__name__} column with the inner type {ClassLabel.__name__}, and column {label_feature} is of type {type(label_feature).__name__}."
```

**Review by mariosasko — COMMENTED** [review]
One last nit and we can merge then

**Comment by lewtun** [comment]
> One last nit and we can merge then

Thanks, done!

**Review by mariosasko — APPROVED** [review]
Looks good now. Thanks!

**Pull request merged**